File: hu/codes/go/chapter_computational_complexity/iteration.go
// File: iteration.go
// Created Time: 2023-08-28
// Author: Reanon (793584285@qq.com)

package chapter_computational_complexity

import "fmt"

/* for ciklus */
func forLoop(n int) int {
	res := 0
	// 1, 2, ..., n-1, n összege
	for i := 1; i <= n; i++ {
		res += i
	}
	return res
}

/* while ciklus */
func whileLoop(n int) int {
	res := 0
	// Feltételváltozó inicializálása
	i := 1
	// 1, 2, ..., n-1, n összege
	for i <= n {
		res += i
		// Feltételváltozó frissítése
		i++
	}
	return res
}

/* while ciklus (két frissítéssel) */
func whileLoopII(n int) int {
	res := 0
	// Feltételváltozó inicializálása
	i := 1
	// 1, 4, 10, ... összege
	for i <= n {
		res += i
		// Feltételváltozó frissítése
		i++
		i *= 2
	}
	return res
}

/* Egymásba ágyazott for ciklus */
func nestedForLoop(n int) string {
	res := ""
	// i = 1, 2, ..., n-1, n ciklus
	for i := 1; i <= n; i++ {
		for j := 1; j <= n; j++ {
			// j = 1, 2, ..., n-1, n ciklus
			res += fmt.Sprintf("(%d, %d), ", i, j)
		}
	}
	return res
}

File: hu/codes/cpp/chapter_computational_complexity/iteration.cpp
/**
 * File: iteration.cpp
 * Created Time: 2023-08-24
 * Author: krahets (krahets@163.com)
 */

#include "../utils/common.hpp"

/* for ciklus */
int forLoop(int n) {
    int res = 0;
    // 1, 2, ..., n-1, n összege
    for (int i = 1; i <= n; ++i) {
        res += i;
    }
    return res;
}

/* while ciklus */
int whileLoop(int n) {
    int res = 0;
    int i = 1; // Feltételváltozó inicializálása
    // 1, 2, ..., n-1, n összege
    while (i <= n) {
        res += i;
        i++; // Feltételváltozó frissítése
    }
    return res;
}

/* while ciklus (két frissítéssel) */
int whileLoopII(int n) {
    int res = 0;
    int i = 1; // Feltételváltozó inicializálása
    // 1, 4, 10, ... összege
    while (i <= n) {
        res += i;
        // Feltételváltozó frissítése
        i++;
        i *= 2;
    }
    return res;
}

/* Egymásba ágyazott for ciklus */
string nestedForLoop(int n) {
    ostringstream res;
    // i = 1, 2, ..., n-1, n ciklus
    for (int i = 1; i <= n; ++i) {
        // j = 1, 2, ..., n-1, n ciklus
        for (int j = 1; j <= n; ++j) {
            res << "(" << i << ", " << j << "), ";
        }
    }
    return res.str();
}

/* Főprogram */
int main() {
    int n = 5;
    int res;

    res = forLoop(n);
    cout << "\nfor loop sum result res = " << res << endl;

    res = whileLoop(n);
    cout << "\nwhile loop sum result res = " << res << endl;

    res = whileLoopII(n);
    cout << "\nwhile loop (two updates) sum result res = " << res << endl;

    string resStr = nestedForLoop(n);
    cout << "\nDouble for loop traversal result " << resStr << endl;

    return 0;
}
